File: software/python/tart/tart/imaging/cpp/gps_time.cpp
/* Copyright (C) Tim Molteno 2008-2010. All rights reserved */

#include "gps_time.h"

#include <cmath>
#include "util.h"

#include <ctime>

/** !\brief Conversion of Julian Day number to GPS week and Seconds of Week reckoned from Saturday midnight
	\param y Year as a four digit number YYYY
	\param m Month 1-12
	\param d Day of Month 1-31
	\param h Hour 0-23
	\param minutes Minutes 0-59
	\param seconds Seconds 0-60

GPS Time is a uniformly counting time scale. GPS time was born at the 1/5/1980 to 1/6/1980 midnight. January 6, 1980 is a Sunday. 
GPS Time counts in weeks and seconds of a week from this instant. The weeks begin at the Saturday/Sunday transition.
The days of the week are numbered, with Sunday being 0, Saturday is day 6. GPS week 0 began at the beginning of the GPS Time Scale. 
Within each week the time is usually denoted as the second of the week. This is a number between 0 and 604,800 ( 60 x 60 x 24 x 7).

*/
gps_time::gps_time(int y, int m, int d, int h, int minutes, double seconds)
{
	if (y < 1980)
	{
	  std::stringstream ss;
	  ss << "Year '" << y << "' must >= 1980"; 
	  throw ss.str();
	}


	struct tm in_time;
	in_time.tm_sec = 0;
	in_time.tm_min = minutes;
	in_time.tm_hour = h;
	in_time.tm_mday = d;
	in_time.tm_mon = m-1; // months since january
	in_time.tm_year = y - 1900;
	
	// This is broken if we use -1 because if there is no local daylight savings time, then this fails.
 	in_time.tm_isdst = -1;

	struct tm gps_tm;
 	gps_tm.tm_sec = 0; // january 6th 1980
 	gps_tm.tm_min = 0;
 	gps_tm.tm_hour = 0;
 	gps_tm.tm_mday = 6; 
 	gps_tm.tm_mon = 0; // months since january
 	gps_tm.tm_year = 80;
 	gps_tm.tm_isdst = -1;
	// time_t gps_t = mktime ( &gps_tm );
	time_t gps_t = timegm ( &gps_tm );
  
	/* call mktime: in_time->tm_yday will be set */
	time_t in_t = timegm ( &in_time );
	m_day_of_year = in_time.tm_yday;
  
	// I don't think that the UTC offset works correctly.
	int utc_offset = gps_time::utc_offset(y,m);
	
	double gps_sec = difftime ( in_t, gps_t ) + seconds + utc_offset;
	double gps_days = floor(gps_sec / sec_per_day);
	double gps_weeks = gps_days / 7.0;

	m_week = floor(gps_weeks);
	sec_of_week._seconds = gps_sec - m_week*7.0*sec_per_day;

#if 0
	std::cout << "gps_sec " << gps_sec << std::endl;
	std::cout << "gps_days " << gps_days << std::endl;
	std::cout << "gps_time " << m_week << ":" << sec_of_week << std::endl;
#endif

}



int gps_time::utc_offset(int y, int m)
{
	int utc_offset = 0;
	if (y == 1981) { if (m >= 7) utc_offset = 1; else utc_offset = 0; }
	if (y == 1982) { if (m >= 7) utc_offset = 2; else utc_offset = 1; }
	if (y == 1983) { if (m >= 7) utc_offset = 3; else utc_offset = 2; }
	if (y == 1984) utc_offset = 3;
	if (y == 1985) { if (m >= 7) utc_offset = 4; else utc_offset = 3; }
	if (y == 1986) utc_offset = 4;
	if (y == 1987) utc_offset = 4;
	if (y == 1988) utc_offset = 5;
	if (y == 1989) utc_offset = 5;
	if (y == 1990) utc_offset = 6;
	if (y == 1991) utc_offset = 7;
	if (y == 1992) { if (m >= 7) utc_offset = 8; else utc_offset = 7; }
	if (y == 1993) { if (m >= 7) utc_offset = 9; else utc_offset = 8; }
	if (y == 1994) { if (m >= 7) utc_offset = 10; else utc_offset = 9; }
	if (y == 1995) utc_offset = 10;
	if (y == 1996) utc_offset = 11;
	if (y == 1997) { if (m >= 7) utc_offset = 12; else utc_offset = 11; }
	if (y == 1998) utc_offset = 12;
	if (y == 1999) utc_offset = 13;
	if (y == 2006) utc_offset = 14;
	if (y >= 2009) utc_offset = 15;
	
	return utc_offset;
}

int gps_time::day_of_week() const
{
  double dow = sec_of_week._seconds / sec_per_day;
  return floor(dow);
}


// difference in seconds between gps_times
double gps_time::diff(const gps_time& b) const
{
  double w = (m_week - b.m_week) * 7 * sec_per_day;
  double s = (sec_of_week - b.sec_of_week);
  return w + s;
}

int gps_time::day_of_year() const
{
  return m_day_of_year;
}


std::ostream& operator<<(std::ostream& os, const gps_time& gt)
{
//	double jd = gt.mm_julian_date;
	os << gt.day_of_year() << ":";
	os << gt.m_week << "-" << gt.day_of_week() << "-" << gt.sec_of_week;
	return os;
}

File: software/containers/ephemeris_server/code/GpsTime.rb
# Copyright (C) Tim Molteno 2008-2013. All rights reserved 

=begin !\brief A GPS time object. http://adn.agi.com/GNSSWeb/Default.aspx

SV Clock and GPS Time stuctures are easy to figure out. GPS Time is kept as Weeks and Seconds from Jan. 5 midnight/Jan. 6 morning of 1980.
This means that we don't have to keep track of UTC leap seconds (or perhaps more accurately, we're not keeping track of them here). This
means that when the date in yyyy, mm, dd format is entered, the conversion to GPS time is done by computing the number of seconds between 
that date and the gps start date.

When converting from gps seconds to UTC we technically should keep track of UTC leap seconds.

TODO. Make the conversion between gps_time and UTC robust to the UTC leap second issue.

=end
require 'time'

class GpsTime  
  SEC_PER_DAY = 86400.0

  # !\brief Construct an invalid gps_time
  #
  def initialize()
  
    @m_week = -9999
    @sec_of_week = nil
    @m_day_of_year = nil
  end

  # !\brief Construct from a UTC year, month, day and hour
  #
  #   !\brief Conversion of Julian Day number to GPS week and Seconds of Week reckoned from Saturday midnight
  # 	\param y Year as a four digit number YYYY
  # 	\param m Month 1-12
  # 	\param d Day of Month 1-31
  # 	\param h Hour 0-23
  # 	\param minutes Minutes 0-59
  # 	\param seconds Seconds 0-60
  # 
  # GPS Time is a uniformly counting time scale. GPS time was born at the 1/5/1980 to 1/6/1980 midnight. January 6, 1980 is a Sunday. 
  # GPS Time counts in weeks and seconds of a week from this instant. The weeks begin at the Saturday/Sunday transition.
  # The days of the week are numbered, with Sunday being 0, Saturday is day 6. GPS week 0 began at the beginning of the GPS Time Scale. 
  # Within each week the time is usually denoted as the second of the week. This is a number between 0 and 604,800 ( 60 x 60 x 24 x 7).
  def initialize(y, m, d, h, minutes, seconds)
    
    throw Exception.new("Year #{y} must >= 1980") if (y < 1980)


    in_time = Time.utc(year=y, month=m, day=d, hour=h, min=minutes, sec_with_frac=0)	
    # This is broken if we use -1 because if there is no local daylight savings time, then this fails.
    #in_time.tm_isdst = -1;

    gps_t = Time.utc(year=1980, month=1, day=6, hour=0, min=0, sec_with_frac=0)

    @m_day_of_year = in_time.yday;
    
    
    gps_sec = ( in_time - gps_t ) + seconds + self.utc_offset(in_time)
    gps_days = (gps_sec / SEC_PER_DAY).floor.to_i
    gps_weeks = gps_days / 7.0;
    
    @m_week = (gps_weeks).floor.to_i
    @sec_of_week = gps_sec - @m_week*7.0*SEC_PER_DAY;
  end

  def self.from_time(utc_date)
    return GpsTime.new(utc_date.year(), utc_date.month(), utc_date.day(), utc_date.hour(), utc_date.min(), utc_date.sec())
  end

  # I don't think that the UTC offset works correctly.
  def utc_offset(t)
    return 18 if (t >= Time.utc(2017, 1, 1))
    return 17 if (t >= Time.utc(2015, 7, 1))
    return 16 if (t >= Time.utc(2012, 7, 1))
    return 15 if (t >= Time.utc(2009, 1, 1))
    return 14 if (t >= Time.utc(2006, 1, 1))
    return 13 if (t >= Time.utc(1999, 1, 1))
    return 12 if (t >= Time.utc(1997, 7, 1))
    return 11 if (t >= Time.utc(1996, 1, 1))
    return 10 if (t >= Time.utc(1994, 7, 1))
    return 9 if (t >= Time.utc(1993, 7, 1))
    return 8 if (t >= Time.utc(1992, 7, 1))
    return 7 if (t >= Time.utc(1991, 1, 1))
    return 6 if (t >= Time.utc(1990, 1, 1))
    return 5 if (t >= Time.utc(1988, 1, 1))
    return 4 if (t >= Time.utc(1985, 7, 1))
    return 3 if (t >= Time.utc(1983, 7, 1))
    return 2 if (t >= Time.utc(1982, 7, 1))
    return 1 if (t >= Time.utc(1981, 7, 1))
    return 0 if (t >= Time.utc(1980, 1, 1))
    return nil
  end
  
  def to_s
    "W:#{@m_week}_SOW:#{@sec_of_week}"
  end
  
  def diff(b)
    #print "GpsTime(#{self}).diff(#{b})"
    w = (@m_week - b.week()) * 7 * SEC_PER_DAY
    s = (@sec_of_week - b.sow())
    return w + s
  end

  def week()
    return @m_week
  end

  def sow()
    return @sec_of_week
  end

  def day_of_year()
    return @m_day_of_year
  end
  
  def day_of_week()
    dow = @sec_of_week / SEC_PER_DAY
    return dow.floor.to_i
  end
  
  def yesterday()
    sow = @sec_of_week - SEC_PER_DAY
    week = @m_week
    if (sow < 0)
      sow += 7*SEC_PER_DAY
      week -= 1
    end
    [week, sow]
  end
  
  def tomorrow()
    sow = @sec_of_week + SEC_PER_DAY
    week = @m_week
    if (sow > 7*SEC_PER_DAY)
      sow -= 7*SEC_PER_DAY
      week += 1
    end
    [week, sow]
  end
  

end

def get_gps_time(date)
  utc_date = Time.parse(date.to_s).utc()
  gt = GpsTime.new(utc_date.year(), utc_date.month(), utc_date.day(), utc_date.hour(), utc_date.min(), utc_date.sec())
end
